atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/OracleShrinkExecutor.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.dbkvs.impl;
+
+import java.sql.SQLException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Stopwatch;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.keyvalue.dbkvs.OracleStandardEditionShrinkConfiguration;
+import com.palantir.atlasdb.keyvalue.dbkvs.OracleTableNameGetter;
+import com.palantir.atlasdb.keyvalue.impl.TableMappingNotFoundException;
+import com.palantir.atlasdb.logging.LoggingArgs;
+import com.palantir.exception.PalantirSqlException;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.UnsafeArg;
+import com.palantir.nexus.db.sql.SqlConnection;
+
+public class OracleShrinkExecutor {
+    private static final Logger log = LoggerFactory.getLogger(OracleShrinkExecutor.class);
+
+    private final SqlConnectionSupplier connectionPool;
+    private final ExecutorService executorService;
+    private final OracleTableNameGetter oracleTableNameGetter;
+    private final OracleStandardEditionShrinkConfiguration shrinkConfig;
+
+    private static Future<Boolean> previousShrinkFuture;
+
+    public OracleShrinkExecutor(
+            SqlConnectionSupplier connectionPool,
+            ExecutorService executorService,
+            OracleTableNameGetter oracleTableNameGetter,
+            OracleStandardEditionShrinkConfiguration shrinkConfig) {
+        this.connectionPool = connectionPool;
+        this.executorService = executorService;
+        this.oracleTableNameGetter = oracleTableNameGetter;
+        this.shrinkConfig = shrinkConfig;
+    }
+
+    public void shrinkAsync(TableReference tableRef) {
+        final long secondsToWaitBeforeShrink = getSecondsToWaitBeforeShrink();
+        previousShrinkFuture = executorService.submit(
```

Comment:
Suppose we have:
At time t: SHRINK A
At time t+5: SHRINK B (will have to wait for 5 minutes before returning - assuming 10 minutes is still not long enough and the operation times out, say, it was a very large table). This blocks the sweep thread for 5 minutes.
At time t+8 SHRINK C (this is not not possible, so this operation will now be at t+13, and will have to wait for 7 minutes, assuming the above)

This is non-ideal, we need an Executor where a task can depend on the result of the previous task, but it checks the result just before it starts executing.